clang/test/CodeGenSYCL/add_ir_attributes_global_variable.cpp

```diff
@@ -0,0 +1,86 @@
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -disable-llvm-passes \
+// RUN:    -triple spir64-unknown-unknown -fsycl-is-device -S \
+// RUN:    -emit-llvm %s -o - | FileCheck %s
+
+// Tests the generation of IR attributes when using
+// __sycl_detail__::add_ir_attributes_global_variable
+
+#include "mock_properties.hpp"
+#include "sycl.hpp"
+
+template <typename... NameValues> struct
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable(NameValues::name...,
+                                                         NameValues::value...)]]
+#endif
+    g {
+  int x;
+
+  constexpr g() : x(1) {}
+  constexpr g(int _x) : x(_x) {}
+};
+
+struct
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable(
+        "Prop11", "Prop12", "Prop13", "Prop14", "Prop15", "Prop16", "Prop17",
+        "Another property string", 2, false, TestEnum::Enum1, nullptr, nullptr, ScopedTestEnum::ScopedEnum2)]]
+#endif
+    h {
+  int x;
+
+  constexpr h() : x(1) {}
+  constexpr h(int _x) : x(_x) {}
+};
+
+template <typename... NameValues> struct
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable(
+        NameValues::name..., "Prop11", "Prop12", "Prop13", "Prop14", "Prop15", "Prop16", "Prop17",
+        NameValues::value..., "Another property string", 2, false, TestEnum::Enum1, nullptr, nullptr, ScopedTestEnum::ScopedEnum2)]]
+#endif
+    gh {
+  int x;
+
+  constexpr gh() : x(1) {}
+  constexpr gh(int _x) : x(_x) {}
+};
+
+template <typename... NameValues> struct
+#ifdef __SYCL_DEVICE_ONLY__
+    [[__sycl_detail__::add_ir_attributes_global_variable(
+        "Prop11", "Prop12", "Prop13", "Prop14", "Prop15", "Prop16", "Prop17", NameValues::name...,
+        "Another property string", 2, false, TestEnum::Enum1, nullptr, nullptr, ScopedTestEnum::ScopedEnum2, NameValues::value...)]]
+#endif
+    hg {
+  int x;
+
+  constexpr hg() : x(1) {}
+  constexpr hg(int _x) : x(_x) {}
+};
+
+constexpr g<prop1, prop2, prop3, prop4, prop5, prop6, prop7> g_v;
+constexpr h h_v;
+constexpr gh<prop1, prop2, prop3, prop4, prop5, prop6, prop7> gh_v;
+constexpr hg<prop1, prop2, prop3, prop4, prop5, prop6, prop7> hg_v;
+
+int main() {
+  sycl::queue q;
+  q.submit([&](sycl::handler &h) {
+    h.single_task<class test_kernel>(
+        [=]() {
+          (void)g_v.x;
```

Comment:
Can you add a case with inheritance? I know you've added AST tests but I think a CodeGen test is useful in this case to ensure correct behavior.

Reply:
Sure! They have been added.